pkg/node/manager/manager_test.go

```diff
@@ -998,3 +1049,125 @@ func (s *managerTestSuite) TestNodeWithSameInternalIP(c *check.C) {
 		c.Errorf("timeout while waiting for NodeAdd() event for node1")
 	}
 }
+
+// TestNodeIpset tests that the ipset entries on the node are updated correctly
+// when a node is updated or removed.
+// It is inspired from TestNode() in manager_test.go.
+func (s *managerTestSuite) TestNodeIpset(c *check.C) {
+	ipsetExpect := func(ipsetMgr *ipsetMock, ip string, expected bool) {
+		setName := iptables.CiliumNodeIpsetV6
+		if v4 := net.ParseIP(ip).To4(); v4 != nil {
+			setName = iptables.CiliumNodeIpsetV4
+		}
+
+		found, err := ipsetContains(ipsetMgr, setName, strings.ToLower(ip))
+		c.Assert(err, check.IsNil)
+
+		if found && !expected {
+			c.Errorf("ipset %s contains IP %s but it should not", setName, ip)
+		}
+		if !found && expected {
+			c.Errorf("ipset %s does not contain expected IP %s", setName, ip)
+		}
+	}
+
+	dp := newSignalNodeHandler()
+	dp.EnableNodeAddEvent = true
+	dp.EnableNodeUpdateEvent = true
+	dp.EnableNodeDeleteEvent = true
+	mngr, err := New(&configMock{
+		// Tunneling and EnableIPv4Masquerade are disabled and enabled,
+		// respectively, to make sure we update the ipset in the
+		// manager (see NodeIpsetNeeded()).
+		Tunneling:            false,
+		EnableIPv4Masquerade: true,
+		// RemoteNodeIdentity is enabled to make sure we don't skip the
+		// ipcache update in NodeUpdated(), and in particular, the
+		// update to nodeIpsAdded. If we skip that part, and an old
+		// node existed, then Nodeupdated() will remove the ipset entry
+		// it just added when calling removenodeFromIPCache().
```

Comment:
Discussing this internally, there seems to be some inconsistent behaviour to address here. This PR is on hold until I've got some time to look into this in the new year.

Reply:
This is now addressed in the second commit.